File: 0415-add-strings/0415-add-strings.cpp
class Solution {
public:
    string addStrings(string num1, string num2) {
         int n1 = num1.length(), n2 = num2.length();
        if (n1 < n2) return addStrings(num2, num1);
        int carry = 0;
        string res;
        for (int i = 0; i < n1; i++) {
            int a = num1[n1-1-i] - '0';
            int b = i < n2 ? num2[n2-1-i] - '0' : 0;
            int sum = a + b + carry;
            res = to_string(sum % 10) + res;
            carry = sum / 10;
        }
        return carry ? to_string(carry) + res : res;
    }
};

File: 0415-add-strings/0415-add-strings.py
class Solution:
    def addStrings(self, num1: str, num2: str) -> str:
        res = []
        carry = 0
        p1, p2 = len(num1) - 1, len(num2) - 1
        while p1 >= 0 or p2 >= 0:
            x1 = ord(num1[p1]) - ord('0') if p1 >= 0 else 0
            x2 = ord(num2[p2]) - ord('0') if p2 >= 0 else 0
            temp = carry + x1 + x2
            res.append(temp % 10)
            carry = temp // 10
            p1 -= 1
            p2 -= 1
        if carry:
            res.append(carry)
        return ''.join(str(x) for x in res[::-1])